File: 2467/solution.py
num_solution = int(input())
solution_list = list(map(int,input().split(' ')))

left = 0
right = len(solution_list)-1
mid = int((left+right)/2)
while (right > left):
    if (solution_list[mid] < 0):
        left = mid+1
    else:
        right = mid
    mid = int((left+right)/2)

answer = []
if (mid==0):
    answer = [solution_list[mid], solution_list[mid+1]]
elif ((mid==len(solution_list)-1) and (solution_list[mid]<=0)):
    answer = [solution_list[mid-1], solution_list[mid]]
else:
    if (mid==len(solution_list)-1):
        min_ph_sum = abs(solution_list[mid-2] + solution_list[mid-1])
        temp_answer = [solution_list[mid-2], solution_list[mid-1]]
    else:
        if (abs(solution_list[mid-2] + solution_list[mid-1]) < abs(solution_list[mid] + solution_list[mid+1])):
            min_ph_sum = abs(solution_list[mid-2] + solution_list[mid-1])
            temp_answer = [solution_list[mid-2], solution_list[mid-1]]
        else:
            min_ph_sum = abs(solution_list[mid] + solution_list[mid+1])
            temp_answer = [solution_list[mid], solution_list[mid+1]]
    for solution in solution_list[:mid]:
        sub_left = mid
        sub_right = len(solution_list)-1
        sub_mid = int((sub_left+sub_right)/2)
        while (sub_right > sub_left):
            if (solution_list[sub_mid] < abs(solution)):
                sub_left = sub_mid+1
            else:
                sub_right = sub_mid
            sub_mid = int((sub_left+sub_right)/2)
        if (sub_mid==mid):
            if (abs(solution_list[sub_mid]+solution) < min_ph_sum):
                min_ph_sum = abs(solution_list[sub_mid]+solution)
                temp_answer = [solution, solution_list[sub_mid]]
        else:
            if (abs(solution_list[sub_mid]+solution) <= abs(solution_list[sub_mid-1]+solution)):
                if (abs(solution_list[sub_mid]+solution) < min_ph_sum):
                    min_ph_sum = abs(solution_list[sub_mid]+solution)
                    temp_answer = [solution, solution_list[sub_mid]]
            else:
                if (abs(solution_list[sub_mid-1]+solution) < min_ph_sum):
                    min_ph_sum = abs(solution_list[sub_mid-1]+solution)
                    temp_answer = [solution, solution_list[sub_mid-1]]
    answer = temp_answer

print(str(answer[0])+" "+str(answer[1]))

File: 2467/solution.rs
use std::io::{self, BufRead};

fn main() {
    let mut solution_num = String::new();
    io::stdin().read_line(&mut solution_num).unwrap();
    let solution_vec: Vec<i32> = io::stdin().lock().lines().next().unwrap().unwrap() 
    .split(' ').map(|s| s.trim())
    .filter(|s| !s.is_empty())        
    .map(|s| s.parse().unwrap())     
    .collect(); 

    let mut left = 0;
    let mut right = solution_vec.len()-1;
    let mut answer: [i32; 2] = [solution_vec[left],solution_vec[right]];
    let mut min_sum = (solution_vec[left]+solution_vec[right]).abs();
    while right > left {
        let ph_sum = solution_vec[left]+solution_vec[right];
        let ph_sum_abs = ph_sum.abs();
        if ph_sum_abs < min_sum {
            answer = [solution_vec[left],solution_vec[right]];
            min_sum = ph_sum_abs;
        }
        if ph_sum < 0 {
            left += 1;
        }
        else if ph_sum > 0 {
            right -= 1;
        }
        else {
            break;
        }
    }
    println!("{} {}", answer[0], answer[1]);
}